test/e2e/framework/self_hosted.go

```diff
@@ -0,0 +1,484 @@
+			images := []string{}
+			for _, image := range input.E2EConfig.Images {
+				images = append(images, fmt.Sprintf("%q", image.Name))
+			}
+			clusterctlVariables["DOCKER_PRELOAD_IMAGES"] = `[` + strings.Join(images, ",") + `]`
+		}
+
+		infrastructureProvider := clusterctl.DefaultInfrastructureProvider
+		if input.InfrastructureProvider != nil {
+			infrastructureProvider = *input.InfrastructureProvider
+		}
+		clusterctl.ApplyClusterTemplateAndWait(ctx, clusterctl.ApplyClusterTemplateAndWaitInput{
+			ClusterProxy: input.BootstrapClusterProxy,
+			ConfigCluster: clusterctl.ConfigClusterInput{
+				LogFolder: filepath.Join(
+					input.ArtifactFolder,
+					"clusters",
+					input.BootstrapClusterProxy.GetName(),
+				),
+				ClusterctlConfigPath:     input.ClusterctlConfigPath,
+				KubeconfigPath:           input.BootstrapClusterProxy.GetKubeconfigPath(),
+				InfrastructureProvider:   infrastructureProvider,
+				Flavor:                   input.Flavor,
+				Namespace:                namespace.Name,
+				ClusterName:              workloadClusterName,
+				KubernetesVersion:        kubernetesVersion,
+				ControlPlaneMachineCount: &controlPlaneMachineCount,
+				WorkerMachineCount:       &workerMachineCount,
+				ClusterctlVariables:      clusterctlVariables,
+			},
+			ControlPlaneWaiters:     input.ControlPlaneWaiters,
+			WaitForClusterIntervals: input.E2EConfig.GetIntervals(specName, "wait-cluster"),
+			WaitForControlPlaneIntervals: input.E2EConfig.GetIntervals(
+				specName,
+				"wait-control-plane",
+			),
+			WaitForMachineDeployments: input.E2EConfig.GetIntervals(
+				specName,
+				"wait-worker-nodes",
+			),
+		}, clusterResources)
+
+		if infrastructureProvider == "docker" {
+			By("Loading CAPI runtime extensions image to the workload cluster")
+			Expect(bootstrap.LoadImagesToKindCluster(ctx, bootstrap.LoadImagesToKindClusterInput{
+				Name:   workloadClusterName,
+				Images: input.E2EConfig.Images,
+			})).To(Succeed())
+		}
+
+		By("Turning the workload cluster into a management cluster")
+		cluster := clusterResources.Cluster
+		// Get a ClusterBroker so we can interact with the workload cluster
+		selfHostedClusterProxy = input.BootstrapClusterProxy.GetWorkloadCluster(
+			ctx,
+			cluster.Namespace,
+			cluster.Name,
+			framework.WithMachineLogCollector(input.BootstrapClusterProxy.GetLogCollector()),
+		)
+
+		capie2e.Byf("Creating a namespace for hosting the %s test spec", specName)
+		selfHostedNamespace, selfHostedCancelWatches = framework.CreateNamespaceAndWatchEvents(
+			ctx,
+			framework.CreateNamespaceAndWatchEventsInput{
+				Creator:             selfHostedClusterProxy.GetClient(),
+				ClientSet:           selfHostedClusterProxy.GetClientSet(),
+				Name:                namespace.Name,
+				LogFolder:           filepath.Join(input.ArtifactFolder, "clusters", "bootstrap"),
+				IgnoreAlreadyExists: true,
+			},
+		)
+
+		By("Loading CAPI runtime extensions image to the workload cluster")
```

Comment:
Is this a mistake? Where is this happening?

Reply:
Oops yeah, this is actually happening on lines 201